Sources/AirbnbSwiftFormatTool/Command.swift

```diff
@@ -0,0 +1,50 @@
+// Created by Cal Stephens on 9/25/23.
+// Copyright © 2023 Airbnb Inc. All rights reserved.
+
+import Foundation
+
+/// A single command line invocation
+struct Command {
+
+  // MARK: Internal
+
+  /// Mock implementation of `Command.run` which can be provided during unit test
+  static var _mockRunCommand: ((Command) -> Int32)?
+
+  let log: Bool
+  let launchPath: String
+  let arguments: [String]
+
+  /// Synchronously runs this command and returns its exit code
+  func run() throws -> Int32 {
+    if let _mockRunCommand = Command._mockRunCommand {
+      return _mockRunCommand(self)
+    }
```

Comment:
In general I think it's preferable if we can avoid having test-only code in the production implementation. I am curious if there is a way for us to avoid the need for `_mockRunCommand`. To that end, I want to propose an idea...

What if `Command` was a value type that encoded all of the information about the command we wanted to run, but didn't actually run the command. Then there could be separate logic in `AirbnbSwiftFormatTool` to run the command.

We could make this more fancy, but at a high level I am thinking that we can initialize `AirbnbSwiftFormatTool` with a way to run the command.

```swift
struct AirbnbSwiftFormatTool {
  init(runCommand: @escaping (Command) -> Int32 = Self.defaultRunCommand(command:)) {
    self.runCommand = runCommand
  }
    
  static func defaultRunCommand(command: Command) -> Int32 {
    // contains an adaptation of the `run()` method on `Command` as it exists now
  }
    
  private let runCommand: (Command) -> Int32
}
```

Then we can pass in an alternate closure for `runCommand` when initializing  `AirbnbSwiftFormatTool` under test.

Granted, we are still adapting production to be a bit more complex, but IMHO the injection point we are creating is cleaner as it's less tied to unit testing. For example, I think this pattern would naturally lend itself to supporting a `--dry-run` option, potentially by injecting a command runner object instead of a closure.

```swift
protocol CommandRunner {
  func runCommand(_ command: Command) -> Int32
}

class DefaultCommandRunner: CommandRunner {

  func runCommand(command: Command) -> Int32 {
    // contains an adaptation of the `run()` method on `Command` as it exists now
  }
}

class DryRunDefaultCommandRunner: CommandRunner {

  func runCommand(command: Command) -> Int32 {
    // prints out the command instead of running it
  }
}

class MockCommandRunner: CommandRunner { … }

struct AirbnbSwiftFormatTool {
  init(
    commandRunner: CommandRunner = DefaultCommandRunner(),
    dryRunCommandRunner: CommandRunner = DryRunDefaultCommandRunner())
  {
    self.commandRunner = commandRunner
    self.dryRunCommandRunner = dryRunCommandRunner
  }
  
  func run() throws {
    // chooses `commandRunner` or `dryRunCommandRunner` based on the `--dry-run` option
  }

  private let commandRunner: CommandRunner
  private let dryRunCommandRunner: CommandRunner
}
```

Then we could initialize the `AirbnbSwiftFormatTool` as follows in a unit test environment.

```swift
let mockCommandRunner = MockCommandRunner()
// configure mockCommandRunner
let mockDryRunCommandRunner = MockCommandRunner()
// configure mockDryRunCommandRunner

let tool = AirbnbSwiftFormatTool(
  commandRunner: mockCommandRunner,
  dryRunCommandRunner: mockDryRunCommandRunner)

// run tool
// assert on state of mockCommandRunner and mockDryRunCommandRunner
```

Reply:
Sure, seems reasonable

Reply:
I took a look and this doesn't appear possible for a few reasons:
 1. We can't call `AirbnbSwiftFormatTool.init` directly. `ArgumentParser` `ParsableCommand` implementations must be initialized via the `static func parse` method in order to correctly set up the instance properties, which are all property wrappers provided by the `ArgumentParser` library.
      - This is not a huge obstacle because we could work around it by just making the `runCommand` property a `var` instead of a `let` and just customize it after calling `init`.
 2. We can't add dependencies like this as instance members of `AirbnbSwiftFormatTool` since `ParsableCommand` must be `Decodable`. I can't think of a way to add a closure or builder-like object as an instance member of `AirbnbSwiftFormatTool` without doing something unusual (like making the builder conform to `Decodable` and having that just populate the default implementation) or manually implementing the `decode` method (adds overhead and doesn't seem like idiomatic usage of the `ArgumentParser` library).

It seems this points us to needing to use a global `static var` closure. 

Do you think it would be preferable to use something like:

```swift
static var runCommand: (Command) -> Int32 = { try $0.run() }
```

which is always called even in production, and can be customized for unit tests, over the current design?

Reply:
Thanks for investigating @calda . What you wrote above makes sense.

I think the static function could work well. Another path could be moving more of the code in `AirbnbSwiftFormatTool` into another object that doesn't have so many limitations (e.g. a `Formatter` object). Then `AirbnbSwiftFormatTool`'s `run()` method could instantiate a `Formatter` and call into that object to do the bulk of the work. In this alternate path, we would write the unit tests on this new object.

I can get onboard with either path.

Reply:
I'll go with the static function approach so we can continue using the existing code structure, which seems to be the idiomatic way to use the `ArgumentParser` library.

Reply:
Here you go: https://github.com/airbnb/swift/pull/242/commits/c13dd242f6ad319117002e0c8a859d453bc7654b